src/paillier.rs

```diff
@@ -80,6 +80,25 @@ impl PaillierEncryptionKey {
         let c = a.modmul(&b, self.0.nn());
         Ok((PaillierCiphertext(c), PaillierNonce(nonce)))
     }
+
+    /// Computes $`a \cdot C_1 + C_2`$, returning the resulting ciphertext.
+    pub(crate) fn multiply_and_add(
+        &self,
+        a: &BigNumber,
+        c1: &PaillierCiphertext,
+        c2: &PaillierCiphertext,
+    ) -> Result<PaillierCiphertext> {
+        // Ciphertext addition is modular multiplication, and ciphertext multiplication
+        // is modular exponentiation.
+        //
+        // Note: We do not use `libpaillier::EncryptionKey::mul`
+        // because it does a check that `0 < a < N`. However, the `a` passed in is usually
+        // in the range `-N/2 < a < N/2` so could fail that check. Instead, we just do the
+        // operations directly.
+        Ok(PaillierCiphertext(
+            modpow(&c1.0, a, self.0.nn()).modmul(&c2.0, self.0.nn()),
```

Comment:
As discussed today, another option here is to move `a` to the right range (namely, `0 < a < N`) before calling into `libpaillier::EncryptionKey::mul`. My worry there though is that if `a` is _not_ in the range we'll have an issue.

Reply:
Where do you see the factor coming from anywhere but plus/minus `q`? I did a quick scan of the paper and couldn't find the other cases, but I certainly may have missed one.

In either case, you're probably right, we should have the range check. My reasoning was that the range of the exponent "shouldn't matter", but it's certainly more misuse-resistant to have that check.

Reply:
In presigning: 
- To make `D`, `gamma` is drawn from `F_q` in Round 1 (Fig 7). See L927 of presign/participant.rs.
- To make `D_hat`, the key share `x_i` is drawn from `F_q` in aux-info Round 1 (Fig 6) and stored for presigning. I didn't find where this is implemented.

Pi-AffG
- To make `A`, `alpha` is drawn from plus-minus `2^(ell + epsilon)` in step 1. See L131 in the code of zkp/piaffg.rs

The rest are all using challenges which are drawn from plus/minus q.

Reply:
And yes, I agree that it technically doesn't matter, but something is wrong in our implementation if we're getting exponents outside that range.

Reply:
Actually, this raises an interesting point. If exponents outside the range means something is wrong with the implementation, should we `panic!` instead of returning `Err`? To me `Result` is best used for errors that a caller can recover from, but if this is unrecoverable (that is, an error that means there's a bug in the program), should we handle things differently?

I guess this could become a broader topic about the codebase, so maybe worth discussing at some point.

Reply:
Ah, yes, interesting. 

In general, for the malicious security case, I would say we should handle it gracefully so we can try to figure out who to blame. But for this particular case, I think we always sampled the exponents ourselves, so either we borked them in storage or we messed up in the implementation. We could throw the `InternalInvariantFailed` error, I suppose.

In the storage case, maybe it _is_ the fault of the calling application (or e.g. if storage is external to this library, maybe it could have been compromised), so we might an informative error, at least. But we could do that with a panic.

I'll put this as a discussion topic for our next weekly meeting.

Reply:
:+1: I'm assuming this isn't a blocker towards merging this in?

Reply:
:+1: I'm assuming this isn't a blocker towards merging this in?

Reply:
Nope, please go ahead!